File: plugins/collector/event/libraries/dpkg.go
// reference: https://github.com/osquery/osquery/tree/aea0d6ef30a5e38f66cca252ebfcebe80d1f231a/osquery/utils/linux/dpkg
package libraries

import (
	"bufio"
	"collector/utils"
	"io"
	"os"
	"strings"
	"time"

	"github.com/chriskaliX/SDK"
	"github.com/chriskaliX/SDK/transport/protocol"
	"github.com/mitchellh/mapstructure"
)

var dpkgFiles = [2]string{"/var/lib/dpkg/status", "/usr/local/var/lib/dpkg/status"}

type Dpkg struct {
	Name     string `mapstructure:"name"`
	Version  string `mapstructure:"version"`
	Source   string `mapstructure:"source"`
	Size     string `mapstructure:"size"`
	Arch     string `mapstructure:"arch"`
	Status   string `mapstructure:"status"`
	Section  string `mapstructure:"section"`
	Priority string `mapstructure:"priority"`
}

func (Dpkg) DataType() int { return 3016 }

func (d *Dpkg) Run(sandbox SDK.ISandbox, sig chan struct{}) (err error) {
	hash := utils.Hash()
	for _, dpkgFile := range dpkgFiles {
		f, err := os.Open(dpkgFile)
		if err != nil {
			continue
		}
		// Limit for dpkg file
		s := bufio.NewScanner(io.LimitReader(f, 25*1024*1024))
		// Look into bufio.ScanLines
		s.Split(func(data []byte, atEOF bool) (advance int, token []byte, err error) {
			// End of the file
			if atEOF && len(data) == 0 {
				return 0, nil, nil
			}
			// Find \nPackage:
			if i := strings.Index(string(data), "\nPackage: "); i >= 0 {
				return i + 1, data[0:i], nil
			}
			// reader is finished, but it's not empty, return the rest of data
			if atEOF {
				return len(data), data, nil
			}
			return 0, nil, nil
		})
		for s.Scan() {
			lines := strings.Split(s.Text(), "\n")
			for _, line := range lines {
				fields := strings.SplitN(line, ": ", 2)
				if len(fields) != 2 {
					continue
				}
				switch fields[0] {
				case "Package":
					d.Name = fields[1]
				case "Version":
					d.Version = fields[1]
				case "Source":
					d.Source = fields[1]
				case "Status":
					d.Status = fields[1]
				case "Architecture":
					d.Arch = fields[1]
				case "Installed-Size":
					d.Size = fields[1]
				case "Section":
					d.Section = fields[1]
				case "Priority":
					d.Priority = fields[1]
				}
			}
			rec := &protocol.Record{
				DataType: int32(d.DataType()),
				Data: &protocol.Payload{
					Fields: make(map[string]string, 8),
				},
			}
			mapstructure.Decode(d, &rec.Data.Fields)
			rec.Data.Fields["package_seq"] = hash
			// Maybe way too many, make the channel chunk
			sandbox.SendRecord(rec)
			d.reset()
			time.Sleep(30 * time.Millisecond)
		}
	}
	return
}

var zeroDpkg = &Dpkg{}

func (d *Dpkg) reset() { *d = *zeroDpkg }

func init() { addEvent(&Dpkg{}) }

File: plugins/collector/src/event/libraries/dpkg.rs
/// Dpkg library collector — data_type 3016.
/// Parses /var/lib/dpkg/status.  Mirrors Go's event/libraries/dpkg.go.
use std::collections::HashMap;
use std::fs;
use std::io::{BufRead, BufReader, Read};
use std::time::Duration;

use anyhow::Result;
use sdk::Client;

use crate::event::{hash, make_record};

const DATA_TYPE: i32 = 3016;

static DPKG_FILES: &[&str] = &[
    "/var/lib/dpkg/status",
    "/usr/local/var/lib/dpkg/status",
];

pub async fn run(client: &mut Client) -> Result<()> {
    let seq = hash();
    for &path in DPKG_FILES {
        let Ok(f) = fs::File::open(path) else { continue };
        let reader = BufReader::new(f.take(25 * 1024 * 1024));
        let records = parse_dpkg(reader);
        for fields in records {
            let mut f = fields;
            f.insert("package_seq".into(), seq.clone());
            let _ = client.send_record(&make_record(DATA_TYPE, f));
            tokio::time::sleep(Duration::from_millis(30)).await;
        }
        // only parse the first file found
        break;
    }
    Ok(())
}

/// Parse dpkg status file into one record per package.
/// Uses paragraph-based parsing: blank line separates packages.
fn parse_dpkg(reader: impl BufRead) -> Vec<HashMap<String, String>> {
    let mut result = Vec::new();
    let mut pkg: HashMap<String, String> = HashMap::new();

    for line in reader.lines().flatten() {
        if line.is_empty() {
            flush_pkg(&mut pkg, &mut result);
            continue;
        }
        // Continuation lines start with whitespace — skip
        if line.starts_with(' ') || line.starts_with('\t') { continue; }
        if let Some((k, v)) = line.split_once(": ") {
            match k {
                "Package"        => { pkg.insert("name".into(),     v.to_owned()); }
                "Version"        => { pkg.insert("version".into(),  v.to_owned()); }
                "Source"         => { pkg.insert("source".into(),   v.to_owned()); }
                "Status"         => { pkg.insert("status".into(),   v.to_owned()); }
                "Architecture"   => { pkg.insert("arch".into(),     v.to_owned()); }
                "Installed-Size" => { pkg.insert("size".into(),     v.to_owned()); }
                "Section"        => { pkg.insert("section".into(),  v.to_owned()); }
                "Priority"       => { pkg.insert("priority".into(), v.to_owned()); }
                _ => {}
            }
        }
    }
    flush_pkg(&mut pkg, &mut result);
    result
}

fn flush_pkg(pkg: &mut HashMap<String, String>, out: &mut Vec<HashMap<String, String>>) {
    if pkg.is_empty() { return; }
    let status = pkg.get("status").cloned().unwrap_or_default();
    if status.contains("installed") {
        out.push(pkg.clone());
    }
    pkg.clear();
}
